File: src/repotask/commands/bug.py
"""`repo-task bug` — the bugfix flow."""

from __future__ import annotations

import json
from typing import Any

import typer
from rich.console import RenderableType
from rich.table import Table

from repotask.commands._context import open_knowledge
from repotask.commands.work import _read_stdin_or_file
from repotask.config import load_config
from repotask.connectors import connector_config, default_system, get_connector
from repotask.errors import RepoTaskError
from repotask.index.families import SYMBOLS_FAMILY
from repotask.kb.slicing import build_pack
from repotask.output import emit, guard
from repotask.workflow import store
from repotask.workflow.analyze import impact_set
from repotask.workflow.dedupe import MIN_SIMILARITY, build_impacts, cluster

app = typer.Typer(no_args_is_help=True)

TRIAGE_CONTRACT = """You have the report, the code it points at, and this project's conventions.
Identify the root cause before proposing a fix, and say which of the impacted symbols you
believe is responsible. If the evidence does not support a single root cause, say what
additional signal you need instead of guessing."""

MERGE_CONTRACT = """These tickets resolve to overlapping code. Shared code is not the same as a
shared root cause: confirm each cluster against the reports before recommending a merge,
name the ticket that should survive, and say what the others add that it would lose."""


@app.command("fetch")
@guard
def bug_fetch(
    ticket: str = typer.Argument(..., help="Bug ticket id."),
    system: str = typer.Option("", "--system", help="Connector name."),
    write: str = typer.Option("", "--write", help="Store this content ('-' reads stdin)."),
    budget: int = typer.Option(0, "--budget", help="Token ceiling for the attached context."),
) -> None:
    """Fetch a bug report together with the code and conventions it implicates."""
    from repotask.commands.work import fetch as fetch_command

    if write or not _has_source(ticket):
        fetch_command(ticket=ticket, system=system, write=write)
        if write or not _has_source(ticket):
            return

    config, kb = open_knowledge(sync=False)
    item = store.work_item(config, ticket)
    report = item.require(store.SOURCE, f"repo-task bug fetch {ticket}")
    symbols = kb.facts(SYMBOLS_FAMILY)
    if not symbols:
        raise RepoTaskError("No symbol index found. Run `repo-task index` first.")

    impact = impact_set(report, symbols)
    pack = build_pack(
        kb,
        config.project.stacks,
        "bugfix",
        budget or config.brief.default_budget,
        [item["path"] for item in impact.files[:20]],
    )
    data = {
        "ticket": ticket,
        "report": report,
        "impact": impact.as_dict(),
        "context": pack.as_dict(),
        "contract": TRIAGE_CONTRACT,
    }
    item.write_json(store.ANALYSIS, {"ticket": ticket, "impact": impact.as_dict()})
    item.touch_meta(kind="bug", analyzed=True)
    emit("bug.fetch", data, lambda payload: payload["contract"])


def _has_source(ticket: str) -> bool:
    config = load_config()
    return store.work_item(config, ticket).source_path.is_file()


@app.command("dedupe")
@guard
def bug_dedupe(
    system: str = typer.Option("", "--system", help="Connector name."),
    query: str = typer.Option("", "--query", help="Provider query selecting the tickets."),
    limit: int = typer.Option(50, "--limit", help="Maximum tickets to compare."),
    threshold: float = typer.Option(
        MIN_SIMILARITY, "--threshold", help="Minimum similarity to group two tickets."
    ),
    tickets_file: str = typer.Option(
        "", "--tickets", help="Read tickets as JSON instead of calling a connector ('-' is stdin)."
    ),
) -> None:
    """Group open bug tickets that resolve to the same code.

    In `mcp` mode the CLI returns the search call for the agent to make; the agent
    pipes the resulting tickets back with `--tickets -`.
    """
    config, kb = open_knowledge(sync=False)
    symbols = kb.facts(SYMBOLS_FAMILY)
    if not symbols:
        raise RepoTaskError("No symbol index found. Run `repo-task index` first.")

    if tickets_file:
        tickets = _parse_tickets(_read_stdin_or_file(tickets_file))
    else:
        name = system or default_system(config)
        settings = connector_config(config, name)
        connector = get_connector(name)
        if settings.mode == "mcp":
            request = connector.mcp_list(settings, query, limit)
            emit(
                "bug.dedupe",
                {
                    "mode": "mcp",
                    "request": request.as_dict(),
                    "nextStep": "repo-task bug dedupe --tickets -",
                    "instruction": (
                        "Call the tool above, then pipe the tickets back as JSON: a list of "
                        "objects with id, title, and body."
                    ),
                },
                lambda data: (
                    f"Call {data['request']['server']}.{data['request']['tool']}, then run: "
                    f"{data['nextStep']}"
                ),
            )
            return
        tickets = [ticket.as_dict() for ticket in connector.list_tickets(settings, query, limit)]

    impacts = build_impacts(tickets, symbols)
    clusters = cluster(impacts, threshold)
    unmatched = sorted(
        {item.ticket_id for item in impacts}
        - {member["id"] for group in clusters for member in group["tickets"]}
    )
    emit(
        "bug.dedupe",
        {
            "mode": "analyze",
            "ticketCount": len(tickets),
            "threshold": threshold,
            "clusters": clusters,
            "unmatched": unmatched,
            "contract": MERGE_CONTRACT,
        },
        _render_clusters,
    )


def _parse_tickets(text: str) -> list[dict[str, str]]:
    try:
        payload = json.loads(text)
    except ValueError as error:
        raise RepoTaskError(f"Could not parse tickets as JSON: {error}") from error
    items = payload.get("tickets", payload) if isinstance(payload, dict) else payload
    if not isinstance(items, list):
        raise RepoTaskError("Tickets must be a JSON list of objects with id, title, and body.")
    tickets = []
    for item in items:
        if not isinstance(item, dict) or not item.get("id"):
            raise RepoTaskError("Every ticket needs at least an 'id'.")
        tickets.append(
            {
                "id": str(item["id"]),
                "title": str(item.get("title", "")),
                "body": str(item.get("body", "")),
            }
        )
    return tickets


def _render_clusters(data: dict[str, Any]) -> RenderableType:
    if data.get("mode") != "analyze":
        return str(data)
    if not data["clusters"]:
        return f"No duplicate candidates among {data['ticketCount']} tickets."
    table = Table(title="Possible duplicate clusters")
    table.add_column("score", justify="right")
    table.add_column("tickets", overflow="fold")
    table.add_column("shared code", overflow="fold")
    for group in data["clusters"]:
        table.add_row(
            f"{group['topScore']:.2f}",
            "\n".join(f"{item['id']} {item['title']}".strip() for item in group["tickets"]),
            "\n".join(group["sharedPaths"]) or "-",
        )
    return table

File: src/commands/bug.rs
//! `repo-task bug` — the bugfix flow.

use anyhow::{bail, Result};
use serde_json::{json, Value};

use crate::commands::work::{fetch, parse_tickets, read_stdin_or_file};
use crate::config;
use crate::connectors::{self, fallback, mcp_json};
use crate::index::families::SYMBOLS_FAMILY;
use crate::kb;
use crate::kb::slicing::build_pack;
use crate::output;
use crate::workflow::analyze::impact_set;
use crate::workflow::dedupe::{build_impacts, cluster};
use crate::workflow::store;

const TRIAGE_CONTRACT: &str =
    "You have the report, the code it points at, and this project's conventions.
Identify the root cause before proposing a fix, and say which of the impacted symbols you
believe is responsible. If the evidence does not support a single root cause, say what
additional signal you need instead of guessing.";

const MERGE_CONTRACT: &str =
    "These tickets resolve to overlapping code. Shared code is not the same as a
shared root cause: confirm each cluster against the reports before recommending a merge,
name the ticket that should survive, and say what the others add that it would lose.";

/// Fetch a bug report together with the code and conventions it implicates.
pub fn bug_fetch(ticket: &str, system: &str, write: &str, budget: Option<usize>) -> Result<()> {
    let config = config::load()?;
    let item = store::work_item(&config, ticket)?;

    // Without a stored report there is nothing to triage: fetch first, and stop there
    // when the connector handed the call to the agent instead of returning the ticket.
    if !write.is_empty() || !item.source_path().is_file() {
        fetch(ticket, system, write)?;
        if !write.is_empty() || !item.source_path().is_file() {
            return Ok(());
        }
    }

    let kb = kb::open(&config, Some(false))?;
    let report = item.require(store::SOURCE, &format!("repo-task bug fetch {ticket}"))?;
    let symbols = kb.facts(SYMBOLS_FAMILY);
    if symbols.is_empty() {
        bail!("No symbol index found. Run `repo-task index` first.");
    }

    let impact = impact_set(&report, &symbols, 25);
    let paths: Vec<String> = impact
        .files
        .iter()
        .take(20)
        .filter_map(|file| file["path"].as_str().map(String::from))
        .collect();
    let pack = build_pack(
        &kb,
        &config.project.stacks,
        "bugfix",
        budget
            .filter(|value| *value > 0)
            .unwrap_or(config.brief.default_budget),
        &paths,
    );

    let data = json!({
        "ticket": ticket,
        "report": report,
        "impact": impact,
        "context": pack,
        "contract": TRIAGE_CONTRACT,
    });
    item.write_json(
        store::ANALYSIS,
        &json!({"ticket": ticket, "impact": impact}),
    )?;
    item.touch_meta(&[("kind", json!("bug")), ("analyzed", json!(true))])?;
    output::emit("bug.fetch", &data, |value| {
        let mut lines = vec![
            format!("Bug {}", value["ticket"].as_str().unwrap_or("")),
            String::new(),
            "Report".to_string(),
            value["report"]
                .as_str()
                .unwrap_or("")
                .trim_end()
                .to_string(),
            String::new(),
            "Impact".to_string(),
        ];
        lines.extend(super::render_impact(&value["impact"]));
        lines.push(String::new());
        lines.push(super::render_pack(&value["context"]));
        lines.push(String::new());
        lines.push(value["contract"].as_str().unwrap_or("").to_string());
        lines.join("\n")
    });
    Ok(())
}

/// Group open bug tickets that resolve to the same code.
pub fn bug_dedupe(
    system: &str,
    query: &str,
    limit: usize,
    threshold: f64,
    tickets_file: &str,
) -> Result<()> {
    let config = config::load()?;
    let kb = kb::open(&config, Some(false))?;
    let symbols = kb.facts(SYMBOLS_FAMILY);
    if symbols.is_empty() {
        bail!("No symbol index found. Run `repo-task index` first.");
    }

    let tickets: Vec<Value> = if !tickets_file.is_empty() {
        parse_tickets(&read_stdin_or_file(tickets_file)?)?
    } else {
        let name = if system.is_empty() {
            connectors::default_system(&config)?
        } else {
            system.to_string()
        };
        let settings = connectors::connector_config(&config, &name)?;
        let connector = connectors::get(&name)?;
        let attempt = fallback::attempt(settings.allows_rest(), settings.allows_mcp(), || {
            connector.list_tickets(&settings, query, limit)
        })?;
        match attempt {
            fallback::Attempt::Rest(tickets) => tickets
                .into_iter()
                .map(|ticket| serde_json::to_value(ticket).unwrap_or_default())
                .collect(),
            fallback::Attempt::FallBack(reason) => {
                let request = connector.mcp_list(&settings, query, limit)?;
                output::warn(format!("Falling back to an MCP call: {reason}"));
                let data = json!({
                    "mode": "mcp",
                    "reason": reason,
                    "request": mcp_json(&request),
                    "nextStep": "repo-task bug dedupe --tickets -",
                    "instruction": "Call the tool above, then pipe the tickets back as JSON: a \
                                    list of objects with id, title, and body.",
                });
                output::emit("bug.dedupe", &data, |value| {
                    format!(
                        "Call {}.{}, then run: {}",
                        value["request"]["server"].as_str().unwrap_or(""),
                        value["request"]["tool"].as_str().unwrap_or(""),
                        value["nextStep"].as_str().unwrap_or(""),
                    )
                });
                return Ok(());
            }
        }
    };

    let impacts = build_impacts(&tickets, &symbols);
    let clusters = cluster(&impacts, threshold);

    let clustered: Vec<String> = clusters
        .iter()
        .filter_map(|group| group["tickets"].as_array())
        .flatten()
        .filter_map(|member| member["id"].as_str().map(String::from))
        .collect();
    let mut unmatched: Vec<String> = impacts
        .iter()
        .map(|item| item.ticket_id.clone())
        .filter(|id| !clustered.contains(id))
        .collect();
    unmatched.sort();
    unmatched.dedup();

    let data = json!({
        "mode": "analyze",
        "ticketCount": tickets.len(),
        "threshold": threshold,
        "clusters": clusters,
        "unmatched": unmatched,
        "contract": MERGE_CONTRACT,
    });
    output::emit("bug.dedupe", &data, render_clusters);
    Ok(())
}

fn render_clusters(value: &Value) -> String {
    let Some(clusters) = value["clusters"].as_array() else {
        return String::new();
    };
    if clusters.is_empty() {
        return format!(
            "No duplicate candidates among {} tickets.",
            value["ticketCount"].as_u64().unwrap_or(0)
        );
    }
    let mut lines = vec!["Possible duplicate clusters".to_string()];
    for group in clusters {
        lines.push(format!(
            "  score {:.2}",
            group["topScore"].as_f64().unwrap_or(0.0)
        ));
        if let Some(tickets) = group["tickets"].as_array() {
            for ticket in tickets {
                lines.push(format!(
                    "    {} {}",
                    ticket["id"].as_str().unwrap_or(""),
                    ticket["title"].as_str().unwrap_or("")
                ));
            }
        }
        if let Some(paths) = group["sharedPaths"].as_array() {
            for path in paths {
                lines.push(format!("      shared: {}", path.as_str().unwrap_or("")));
            }
        }
    }
    lines.join("\n")
}
